writer.go

```diff
@@ -1232,20 +1235,29 @@ func (c *writerColumn) flushFilterPages() error {
 		index:              int16(c.bufferIndex),
 	}
 
-	rbuf, pool := getBufioReader(nil, 1024)
+	var pageReader io.Reader = c.pageBuffer
+	if _, err = c.pageBuffer.Seek(0, io.SeekStart); err != nil {
+		return err
+	}
+
+	if _, ok := pageReader.(*os.File); ok {
+		rbuf, pool := getBufioReader(pageReader, 1024)
+		defer func() {
+			putBufioReader(rbuf, pool)
+		}()
+		pageReader = rbuf
+	}
+
 	pbuf := (*buffer)(nil)
 	defer func() {
-		putBufioReader(rbuf, pool)
 		if pbuf != nil {
 			pbuf.unref()
 		}
 	}()
 
-	decoder := thrift.NewDecoder(c.header.protocol.NewReader(rbuf))
-
-	for _, p := range c.pages {
-		rbuf.Reset(p)
+	decoder := thrift.NewDecoder(c.header.protocol.NewReader(pageReader))
 
+	for i := 0; i < c.numPages; i++ {
```

Comment:
This is the only usage of numPages, and it doesn't _seem_ like it should be necessary - instead it seems like we should be able to just iterate through the whole pageBuffer and stop when we finish. However, I'm not sure the best way to construct that loop and handle error cases.

Reply:
In theory we should be able to copy raw bytes if pages are already properly constructed into the write buffer.

Reply:
Can you elaborate on this? I don't follow what you're getting at.

Reply:
What I was thinking is that the write buffer already has encoded and compressed pages, so what is stopping us from just copying the write buffer as is into the target file/writer?

Reply:
That's what this PR does on line 868. However, here we're populating the bloom filter, which requires decoding the pages so that the right values are added.

Reply:
Got it, this makes sense to me 👍